tensorflow/contrib/tensorrt/segment/segment.cc

```diff
@@ -32,74 +32,92 @@ namespace tensorflow {
+  int dst_input() const { return dst_port_; }
+  int id() const { return id_; }
+  bool IsControlEdge() const { return control_; }
+  ~SimpleEdge() {}
+
+ private:
+  int id_;
+  SimpleNode* src_;
+  int src_port_;
+  SimpleNode* dst_;
+  int dst_port_;
+  bool control_;
+};
+class SimpleNode {
+ public:
+  SimpleNode(const tensorflow::Node* node, const int id);
+  const std::vector<SimpleEdge*>& in_edges() const { return in_edges_; };
+  const std::vector<SimpleEdge*>& out_edges() const { return out_edges_; };
+  std::vector<SimpleNode*> in_nodes() const {
+    std::vector<SimpleNode*> res;
+    res.reserve(in_edges_.size());
+    for (const auto e : in_edges_) {
+      if (e) res.push_back(e->src());
     }
+    return res;
   }
-  return false;
-}
-
-bool CanContractEdge(const SimpleEdge* edge, const SimpleGraph* graph) {
-  const auto src = edge->src();
-  const auto dst = edge->dst();
-
-  // Can't contract edge if doing so would cause a cycle in the
-  // graph. So, if there is a directed path from 'src' to 'dst', other
-  // than 'edge' (or any other direct edge from 'src' to 'dst'), then
-  // combining 'src' and 'dst' will cause a cycle along that path.
-  //
-  // In practice, to avoid modifying the graph and to take advantage
-  // of existing graph functions, we perform an equivalent.
-  //   1. Get all nodes incoming to 'dst', excluding 'src'
-  //   2. Reverse DFS from those nodes
-  //   3. If reverse DFS reaches 'src' then we have a cycle
-  std::vector<SimpleNode*> dfs_start_nodes;
-  for (SimpleNode* node : dst->in_nodes()) {
-    if (node != src) {
-      dfs_start_nodes.push_back(node);
-    }
+  const string& name() const { return node_->name(); }
+  const tensorflow::Node* tf_node() const { return node_; }
+  int id() const { return id_; }
+
+ private:
+  const tensorflow::Node* node_;
+  std::vector<SimpleEdge*> in_edges_;
+  std::vector<SimpleEdge*> out_edges_;
+  int id_;
+
+  friend class SimpleGraph;
+};
+
+class SimpleGraph {
+ public:
+  SimpleGraph(const tensorflow::Graph* g);
+  void AddControlEdge(SimpleNode* src, SimpleNode* dst);
+  void AddEdge(SimpleNode* src, int out_port, SimpleNode* dst, int in_port);
+  void RemoveEdge(const SimpleEdge*);
+  SimpleNode* FindNodeId(int node_id) {
+    if (node_id < 0 || node_id > (int)nodes_.size()) return nullptr;
+    return nodes_[node_id];
+  }
+  ~SimpleGraph();
+  int num_node_ids() const { return nodes_.size(); }
+  const SimpleNode* source_node() const {
+    return nodes_[tensorflow::Graph::kSourceId];
+  }
+  const SimpleNode* sink_node() const {
+    return nodes_[tensorflow::Graph::kSinkId];
   }
 
-  bool is_cycle = CheckCycles(graph, src, dfs_start_nodes);
-  return !is_cycle;
-}
-}  // namespace
+ private:
+  const tensorflow::Graph* g_;
+  std::vector<SimpleNode*> nodes_;
+  std::vector<SimpleEdge*> edges_;
+  // edge_ids_ and node_ids_ contain freed indices.
+  std::set<int> free_edge_ids_;
+  std::set<int> free_node_ids_;
```

Comment:
This is not used, remove?

Reply:
I would like to keep it. I believe I may need it soon for dyn ops